add howto on plugins (#4078)
aiidateam/aiida-core: docs/source/topics/plugins.rst

.. _topics:plugins:maynot:

What a plugin should not do
===========================

An AiiDA plugin should not:

* Change the database schema AiiDA uses
* Use protected functions, methods or classes of AiiDA (those starting with an underscore ``_``)
* Monkey patch anything within the ``aiida`` namespace (or the namespace itself)

Failure to comply will likely prevent your plugin from being listed on the official `AiiDA plugin registry <registry>`_.

If you find yourself tempted to do any of the above, please open an issue on the `AiiDA repository <core>`_ and explain why.
We will advise on how to proceed.

.. _core: https://github.com/aiidateam/aiida-core
.. _registry: https://github.com/aiidateam/aiida-registry

.. _topics:plugins:entrypoints:

What is an entry point?
=======================

The ``setuptools`` package (used by ``pip``) has a feature called `entry points`_, which allows to associate a string (the entry point *identifier*) with any python object defined inside a python package.
Entry points are defined in the ``setup.py`` file, for example::

      ...
      entry_points={
         "aiida.data": [
             # entry point = path.to.python.object

>              "mycode.mydat = aiida_mycode.data.mydat:MyData",

Author: No idea... I agree that `mydata` would be better

   class MyScheduler(Scheduler):
      ...

Usage: The scheduler is used in the familiar way by entering 'myscheduler' as the scheduler option when setting up a computer.

``aiida.transports``
--------------------

``aiida-core`` ships with two modes of transporting files and folders to remote computers: ``ssh`` and ``local`` (stub for when the remote computer is actually the same).
We recommend naming the plugin package after the mode of transport (e.g. ``aiida-mytransport``), so that the entry point name can simply equal the name of the transport:

Spec::

   entry_points={
      "aiida.transports": [
         "mytransport = aiida_mytransport.mytransport:MyTransport"
      ]
   }

``aiida_mytransport/mytransport.py``::

   from aiida.transports import Transport
   class MyTransport(Transport):
      ...

Usage::

   from aiida.plugins import TransportFactory
   transport = TransportFactory('mytransport')

When setting up a new computer, specify ``mytransport`` as the transport mode.

.. _topics:plugins:testfixtures:

Plugin test fixtures
====================

One concern when running tests for AiiDA plugins is to separate the test environment from your production environment.

> Depending on the kind of test, each should even be run against a fresh AiiDA database.

Reviewer: ```suggestion
Typically tests should be run against an empty AiiDA database.
```
